🆕 (android) Add a page detailing the camera orientations (#58)
chromeos/chromeos.dev: pages/en/android/camera-orientation.md

---
title: Camera orientations
metadesc: An in-depth guide on how to handle orientations in Android camera apps.
date: 2020-09-23
weight: 1
tags:

>   - camera orientation
>   - large screens

Reviewer: ```suggestion
  - camera
```

Please remove the "large screens" tag as this article doesn't appear to be about utilizing extra space for larger screens. If that's not correct, feel free to keep it. Camera in next comment

resources:
  - title: CameraX Preview Documentation
    url: https://developer.android.com/reference/androidx/camera/core/Preview
---

If your Android app uses cameras, there are some special considerations when handling orientations. This article presumes you understand the basic concepts of Android camera2 API. You can read our [blog post](https://medium.com/google-developers/detecting-camera-features-with-camera2-61675bb7d1bf) or [summary](https://developer.android.com/reference/android/hardware/camera2/package-summary) for an overview of camera2. We also recommend that you try writing a simple camera app first before approaching the article.

## Background

Handling orientations in Android camera apps is tricky and needs to take the following factors into consideration:

> 1. Natural orientation: The display orientation when the device is in the “normal” position for the device design - usually portrait orientation for mobile phones and landscape orientation for laptops.
> 2. Sensor orientation: The orientation of the sensor physically mounted on the device.
> 3. Display rotation: How much the device is physically rotated from the natural orientation.
> 4. Viewfinder size: The size of the viewfinder used for displaying the camera preview.
> 5. Image size output by the camera.

Reviewer: These don't read like an ordered list to me, but rather like a group of things that could be in any order. If that's true, can we make it an unordered list? If not, can it be rewritten so the items are clearly needed in a specific order?

Thanks!
```suggestion
* Natural orientation: The display orientation when the device is in the “normal” position for the device design - usually portrait orientation for mobile phones and landscape orientation for laptops.
* Sensor orientation: The orientation of the sensor physically mounted on the device.
* Display rotation: How much the device is physically rotated from the natural orientation.
&* Viewfinder size: The size of the viewfinder used for displaying the camera preview.
* Image size output by the camera.
```

Author: There is an extra ampersand that I cannot take out here. 

> &* Viewfinder size

Reviewer: Removed it, thanks for catching!

Author: Awesome. Thanks for the suggestion here :)

| 270              | ![A landscape-shaped image with the bugdroid's head pointed upwards](/images/android/camera-orientation/surface_view_display_rotation_270_phone.png) | ![A portrait-shaped image with the bugdroid's head pointed upwards](/images/android/camera-orientation/surface_view_display_rotation_270_laptop.png)  |

#### Layout

As you can see, SurfaceView already handles some of the tricky things for us. But you now need to consider the size of the viewfinder, or how big you want your preview on the screen. SurfaceView automatically scales the source buffer to fit its dimensions. You need to make sure the aspect ratio of the viewfinder is identical to that of the sourcebuffer. For instance, If you try to fit a portrait-shaped preview into a landscape-shaped SurfaceView, you’ll get something distorted like this:

![Image illustration showing a stretched bugdroid as a result of fitting a portrait-shaped preview into a landscape-shaped viewfinder](/images/android/camera-orientation/surface_view_stretched_scene.png)

> You generally want **the aspect ratio (i.e., width/height) of the viewfinder to be identical to the aspect ratio of the source**. If you don’t want to clip the image in the viewfinder - cutting off some of the pixels to fix the display, there are 2 cases to consider:
> 
> ##### 1. `aspectRatioActivity > aspectRatioSource`

Reviewer: ```suggestion
You generally want **the aspect ratio (i.e., width/height) of the viewfinder to be identical to the aspect ratio of the source**. If you don’t want to clip the image in the viewfinder - cutting off some of the pixels to fix the display, there are 2 cases to consider, when `aspectRatioActivity` is greater than `aspectRatioSource` and when it's less than or equal to `aspectRatioSource`

##### `aspectRatioActivity > aspectRatioSource`
```

You can think of the case as the activity being “wider”. Below we consider an example where you have a 16:9 activity and a 4:3 source.

```
aspectRatioActivity = 16/9 ≈ 1.78
aspectRatioSource = 4/3 ≈ 1.33
```

First you want your viewfinder to be 4:3 as well. Then you want to fit the source and viewfinder into the activity like so:

![Illustration of an activity whose aspect ratio is greater than the aspect ratio of the viewfinder inside](/images/android/camera-orientation/surface_view_aspect_ratio_activity_bigger.png)

In this case, you should make the height of the viewfinder match the height of the activity while making the aspect ratio of the viewfinder identical to the aspect ratio of the source. Very simple pseudo code to demonstrate:

```
viewfinderHeight = activityHeight;
viewfinderWidth = activityHeight * aspectRatioSource;
```

> ##### 2. `aspectRatioActivity ≤ aspectRatioSource`

Reviewer: ```suggestion
##### `aspectRatioActivity ≤ aspectRatioSource`
```

Before calling setDefaultBufferSize, it’s important that you **also set the size of the TextureView to be `previewWidth × previewHeight`** with [View#setLayoutParams(android.view.ViewGroup.LayoutParams)](<https://developer.android.com/reference/android/view/View#setLayoutParams(android.view.ViewGroup.LayoutParams)>). The reason for this is that TextureView calls `SurfaceTexture#setDefaultBufferSize(int width, height)` with its measured width and height. If the size of the TextureView is not explicitly set beforehand, it can cause a race condition. This is mitigated by explicitly setting the size of the TextureView first.

Now the TextureView may not match the dimensions of the source. In the case of phones, the source is portrait-shaped, yet the TextureView is landscape-shaped owing to the layoutParams you just set. This would result in a stretched previews, as illustrated below,

![Image illustration of a portrait-shaped preview stretched to fit inside a TextureView of the same size of the preview size chosen](/images/android/camera-orientation/texture_view_step_1_identical_size.png)

> ##### 2. Scale the potentially stretched TextureView back to the original dimensions of the preview

Reviewer: Please ensure these items in the process are consistently headers or just bold text in a paragraph, and please don't include numbers if they're headers.

Author: Here we're trying to show a 3-step process, so it's procedural.

The problem here is each of the 3 steps needs quite a bit of paragraph, and since it's procedural, I think it makes sense to include numbers to indicate which step we're in.

Let me think how this can be dealt with within the constraints of Markdown.

Author: Corrected :) Thanks!

##### 3. Rotate the preview by `displayRotation` counterclockwise

As previously mentioned, you should rotate the preview by `displayRotation` counterclockwise to compensate for display rotation.

You can do this by [View#setRotation(float)](<https://developer.android.com/reference/android/view/View#setRotation(float)>)

- setRotation(`-displayRotation`), since it does a clockwise rotation.

![Image illustration showing the procedure of the preview being rotated to match to the display orientation of the device](/images/android/camera-orientation/texture_view_step_3_rotate.png)

##### Sample

- [PreviewView](https://developer.android.com/reference/androidx/camera/view/PreviewView) from [camerax](https://developer.android.com/training/camerax) in [Jetpack](https://developer.android.com/android/jetpack) handles TextureView layout as described above. It configures the transformation with [PreviewCorrector](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-master-dev:camera/camera-view/src/main/java/androidx/camera/view/preview/transform/PreviewCorrector.java).

> ##### Note
> 
> If you have previously used transformation matrix for TextureView in your code, the preview may not look right on a naturally-landscape device like Chromebooks. Likely your transformation matrix incorrectly assumes the sensor orientation to be 90 or 270 degrees. You may refer to this [commit](https://github.com/android/camera-samples/commit/3d1a254eb018a51ff39ae78d39a9e9e7942a027b) on GitHub for a workaround, but we highly recommend that you migrate your app to use the method described here instead.

Reviewer: ```suggestion
!!! aside.message--note
**Note:** If you have previously used transformation matrix for TextureView in your code, the preview may not look right on a naturally-landscape device like Chromebooks. Likely your transformation matrix incorrectly assumes the sensor orientation to be 90 or 270 degrees. You may refer to this [commit](https://github.com/android/camera-samples/commit/3d1a254eb018a51ff39ae78d39a9e9e7942a027b) on GitHub for a workaround, but we highly recommend that you migrate your app to use the method described here instead.
!!!
```

We have a specific set of markdown fences to render this content as a note that should be used here.
